src/dsimhost_fpga.py

```diff
@@ -0,0 +1,176 @@
+import logging
+import time
+
+from casperfpga.attribute_container import AttributeContainer
+from casperfpga import tengbe
+
+from corr2.host_fpga import FpgaHost
+
+LOGGER = logging.getLogger(__name__)
+
+def get_prefixed_name(prefix, string):
+    if not string.startswith(prefix):
+        return None
+    if string.startswith(prefix+'_'):
+        return string[len(prefix)+1:]
+    else:
+        return string[len(prefix):]
+
+class Source(object):
+    def __init__(self, register, name):
+        self.register = register
+        self.name = name
+
+class SineSource(Source):
+    def set(self, scale=None, frequency=None):
+        # TODO work in 'human' units such as Hz or Volts?
+        self.register.write(scale=scale, frequency=frequency)
+
+class NoiseSource(Source):
+    def set(self, scale=None, frequency=None):
+        # TODO work in 'human' units such W / Hz?
+        self.register.write(scale=scale, frequency=frequency)
+
+class Output(object):
+    def __init__(self, name, scale_register, control_register):
+        self.name = name
+        self.scale_register = scale_register
+        self.control_register = control_register
+
+    def select_output(self, output_type):
+        tgv_select = 'tvg_select' + self.name
+        if output_type == 'test_vectors':
+            self.control_register.write(**{tgv_select: 0})
+        elif output_type == 'signal':
+            self.control_register.write(**{tgv_select: 1})
+        else:
+            raise ValueError(
+                'Valid output_type values: "test_vectors" and "signal"')
+
+    def scale_output(self, scale):
+        # TODO normalized scale to take values between 0 and 1?
+        self.scale_register.write(scale=scale)
+
+class FpgaDsimHost(FpgaHost):
+    def __init__(self, host, katcp_port=7147, boffile=None, connect=True, config=None):
+        if config:
+            if boffile:
+                raise ValueError('Cannot specify "boffile" parameter if config is used')
+            boffile = config['bitstream']
+        self.config = config
+        FpgaHost.__init__(
+            self, host, katcp_port=katcp_port, boffile=boffile, connect=connect)
+        self.boffile = boffile
+        self.sine_sources = AttributeContainer()
+        self.noise_sources = AttributeContainer()
+        self.outputs = AttributeContainer()
+
+    def get_system_information(self, filename=None, fpg_info=None):
+        """Get system information and build D-engine sources"""
+        FpgaHost.get_system_information(self, filename=filename, fpg_info=fpg_info)
+        self.sine_sources.clear()
```

Comment:
AttributeContrainer doesn't have a clear() method - although it probably should. Have you added one?